paimon-core/src/main/java/org/apache/paimon/metrics/commit/CommitMetrics.java

```diff
@@ -0,0 +1,246 @@
+import org.apache.paimon.metrics.DescriptiveStatisticsHistogram;
+import org.apache.paimon.metrics.Histogram;
+import org.apache.paimon.metrics.MetricGroup;
+import org.apache.paimon.metrics.groups.GenericMetricGroup;
+import org.apache.paimon.utils.FileStorePathFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/** Metrics to measure a commit. */
+public class CommitMetrics {
+    private static Logger log = LoggerFactory.getLogger(CommitMetrics.class);
+    private static final int HISTOGRAM_WINDOW_SIZE = 10_000;
+    protected static final String GROUP_NAME = "commit";
+
+    private final MetricGroup genericMetricGroup;
+
+    private final FileStorePathFactory pathFactory;
+    private final FileIO fileIO;
+
+    private long initTableFilesCount = 0;
+    private long initChangelogFilesCount = 0;
+
+    public CommitMetrics(FileStorePathFactory pathFactory, FileIO fileIO) {
+        this.genericMetricGroup =
+                GenericMetricGroup.createGenericMetricGroup(
+                        pathFactory.root().getName(), GROUP_NAME);
+        this.pathFactory = pathFactory;
+        this.fileIO = fileIO;
+        initDataFilesCount();
+        registerGenericCommitMetrics();
+    }
+
+    @VisibleForTesting
+    void initDataFilesCount() {
+        try {
+            List<Path> dirs =
+                    Arrays.stream(fileIO.listStatus(pathFactory.root()))
+                            .map(f -> f.getPath())
+                            .collect(Collectors.toList());
+            boolean hasPartition = true;
+            for (Path dir : dirs) {
+                if (dir.getName().startsWith("bucket-")) {
+                    hasPartition = false;
+                    break;
+                }
+            }
+            if (hasPartition) {
+                List<Path> buckets = new ArrayList<>();
+                for (Path dir : dirs) {
+                    FileStatus[] fileStatuses = fileIO.listStatus(dir);
+                    buckets.addAll(
+                            Arrays.stream(fileStatuses)
+                                    .filter(
+                                            f ->
+                                                    f.isDir()
+                                                            && f.getPath()
+                                                                    .getName()
+                                                                    .startsWith("bucket-"))
+                                    .map(f -> f.getPath())
+                                    .collect(Collectors.toList()));
+                }
+                for (Path bucket : buckets) {
+                    FileStatus[] fileStatuses = fileIO.listStatus(bucket);
+                    accFilesCount(fileStatuses);
+                }
```

Comment:
For some filesystem, especially object store, listing directory is super slow. Never enumerate table files by walking through the directory tree structure. This code will also be slow if the user has accumulated hundreds of partitions.